### PR TITLE
Add support for human readable names

### PR DESCRIPTION
**Pull request by neilcx**
- Enable human readable names for BOSH created VMs
- Add Unit Test

[153524141]https://www.pivotaltracker.com/n/projects/2110693/stories/153524141

### CODE REVIEW

File: src/vsphere_cpi/lib/cloud/vsphere/cloud.rb [hunk 1]
```diff
@@ -295,6 +295,8 @@ def create_vm(agent_id, stemcell_cid, vm_type, networks_spec, existing_disk_cids
             },
             global_clusters: @datacenter.clusters,
             disk_configurations: disk_configurations(vm_type,  existing_disk_cids),
+            human_readble_name_info: get_name_info_from_bosh_env(environment),
```
Comment (ktchen14) [inline src/vsphere_cpi/lib/cloud/vsphere/cloud.rb]: Typo

File: src/vsphere_cpi/lib/cloud/vsphere/cloud.rb [hunk 2]
```diff
@@ -856,5 +858,34 @@ def disk_configurations(vm_type, existing_disk_cids=[])
...
+    # The bosh is guaranteed to give a environment with no error
+    # which means in the groups array , there are always
+    # director-name, deployment-name and job-name in order
+
+    def get_name_info_from_bosh_env(environment)
+      name_info = {}
+      if environment.nil? || environment.empty?
+        return name_info
+      end
+      instance_group_name = environment.fetch('bosh', {}).fetch('groups', {}).fetch(2,"")
```
Comment (ktchen14) [inline src/vsphere_cpi/lib/cloud/vsphere/cloud.rb]: You can use `Hash#dig` here.

http://mitrev.net/ruby/2015/11/13/the-operator-in-ruby/

File: src/vsphere_cpi/lib/cloud/vsphere/vm_config.rb [hunk 3]
```diff
@@ -52,10 +92,58 @@ def agent_env
...
 
+    # unit test passed
+    # Question here
+    def human_readable_name_enabled?
+      @manifest_params[:enable_human_readable_name]
+    end
+
+    # unit test passed
+    def human_readable_name_info
+      @manifest_params[:human_readable_name_info] || {}
```
Comment (ktchen14) [inline src/vsphere_cpi/lib/cloud/vsphere/vm_config.rb]: `Hash#fetch`

File: src/vsphere_cpi/spec/unit/cloud/vsphere/vm_config_spec.rb [hunk 4]
```diff
@@ -48,15 +50,44 @@ def fake_datastore(name, free_space, mob = nil)
...
+        context 'when human readable name enabled  is set to true' do
+          let(:input){ { human_readable_name_info: { 'instance_group_name' => 'fake-instance-group-name', 'deployment_name'=>'fake-deployment-name'} } }
+          context 'when human readable name info has both instance group name and deployment name' do
+            it 'returns a valid human readable name' do
+              allow(vm_config).to receive(:human_readable_name_enabled?).and_return(true)
+              name = vm_config.name
+              expect(name).to start_with( 'fake-instance-group-name_fake-deployment-name' )
+            end
+          end
+        end
```
Comment (EleanorRigby) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/vm_config_spec.rb]: What about the cases when `name_info` does not have all the information.

File: src/vsphere_cpi/spec/unit/cloud/vsphere/vm_config_spec.rb [hunk 5]
```diff
@@ -153,13 +184,111 @@ def fake_datastore(name, free_space, mob = nil)
...
+
     describe '#networks_spec' do
       let(:input) { { networks_spec: { 'fake-key' => 'fake-value' } } }
       it 'returns the provided networks_spec' do
         expect(vm_config.networks_spec).to eq({ 'fake-key' => 'fake-value' })
       end
     end
 
+
+    # currently only works for english
```
Comment (EleanorRigby) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/vm_config_spec.rb]: Remove this.

File: src/vsphere_cpi/spec/unit/cloud/vsphere/vm_config_spec.rb [hunk 6]
```diff
@@ -153,13 +184,111 @@ def fake_datastore(name, free_space, mob = nil)
...
+    describe '#generate_human_readable_name' do
+
```
Comment (EleanorRigby) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/vm_config_spec.rb]: Remove line break

File: src/vsphere_cpi/lib/cloud/vsphere/cloud.rb [hunk 7]
```diff
@@ -856,5 +858,34 @@ def disk_configurations(vm_type, existing_disk_cids=[])
...
+    # return name info as a hash
+    # if  there are both instance group name and deployment name
+    #   return a valid hash name info table
+    # else return a empty hash
+    #
+    # The bosh is guaranteed to give a environment with no error
+    # which means in the groups array , there are always
+    # director-name, deployment-name and job-name in order
+
+    def get_name_info_from_bosh_env(environment)
```
[the remark this answers is not in the document]
Comment (EleanorRigby) [reply src/vsphere_cpi/lib/cloud/vsphere/cloud.rb]: https://gist.github.com/chetan/1827484

Comment (ktchen14) [inline src/vsphere_cpi/lib/cloud/vsphere/cloud.rb]: This function looks overcomplicated. Can be essentially:

```ruby
return {} if environment.nil?

instance_group_name = environment.dig('bosh', 'groups', 2)
deployment_name = environment.dig('bosh', 'groups', 1)

return {} if instance_group_name.nil? or deployment_name.nil?

{ instance_group_name: instance_group_name, deployment_name: deployment_name }
```

**Comment by neilcx** [comment]
Close this PR, using a new one 'https://github.com/cloudfoundry/bosh-vsphere-cpi-release/pull/205'

**Pull request closed without merging**